packages/flutter/lib/src/material/carousel.dart

```diff
@@ -490,6 +634,505 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
+
+  // Given an index, this method returns the layout offset for the item. The `index`
+  // is firstly compared to `_firstVisibleItemIndex` and compute the distance
+  // between them, then compute all the current extents for items that are located
+  // in front.
+  @override
+  double indexToLayoutOffset(
+    @Deprecated(
+      'The itemExtent is already available within the scope of this function. '
+      'This feature was deprecated after v3.20.0-7.0.pre.'
+    )
+    double itemExtent,
+    int index,
+  ) {
+    if (index == _firstVisibleItemIndex) {
+      if (_distanceToLeadingEdge <= effectiveShrinkExtent) {
+        return constraints.scrollOffset - effectiveShrinkExtent + _distanceToLeadingEdge;
+      }
+      return constraints.scrollOffset;
+    }
+    double visibleItemsTotalExtent = _distanceToLeadingEdge;
+    for (int i = _firstVisibleItemIndex + 1; i < index; i++) {
+      visibleItemsTotalExtent += _buildItemExtent(i, _currentLayoutDimensions);
+    }
+    return constraints.scrollOffset + visibleItemsTotalExtent;
+  }
+
+  @override
+  int getMinChildIndexForScrollOffset(
+    double scrollOffset,
+    @Deprecated(
+      'The itemExtent is already available within the scope of this function. '
+      'This feature was deprecated after v3.20.0-7.0.pre.'
+    )
+    double itemExtent,
+  ) {
+    return math.max(_firstVisibleItemIndex, 0);
+  }
+
+  @override
+  int getMaxChildIndexForScrollOffset(
+    double scrollOffset,
+    @Deprecated(
+      'The itemExtent is already available within the scope of this function. '
+      'This feature was deprecated after v3.20.0-7.0.pre.'
+    )
+    double itemExtent,
+  ) {
+    final int? childCount = childManager.estimatedChildCount;
+    if (childCount != null) {
+      double visibleItemsTotalExtent = _distanceToLeadingEdge;
+      for (int i = _firstVisibleItemIndex + 1; i < childCount; i++) {
+        visibleItemsTotalExtent += _buildItemExtent(i, _currentLayoutDimensions);
+        if (visibleItemsTotalExtent >= constraints.viewportMainAxisExtent) {
+          return i;
+        }
+      }
+    }
+    return childCount ?? 0;
+  }
+
+  @override
+  double computeMaxScrollOffset(
+    SliverConstraints constraints,
+    @Deprecated(
+      'The itemExtent is already available within the scope of this function. '
+      'This feature was deprecated after v3.20.0-7.0.pre.'
+    )
+    double itemExtent,
+  ) {
+    return childManager.childCount * maxChildExtent;
+  }
+
+  BoxConstraints _getChildConstraints(int index) {
+    final double extent = itemExtentBuilder!(index, _currentLayoutDimensions)!;
+    return constraints.asBoxConstraints(
+      minExtent: extent,
+      maxExtent: extent,
+    );
+  }
+
+  // This method is mostly the same as its parent class [RenderSliverFixedExtentList].
+  // The difference is when we allow some space before the leading items or after
+  // the trailing items with smaller weights, we leave extra scroll offset.
```

Comment:
Rather than copying most of this logic, can we add a spacing value/getter/method on the super class? If this is feasible, we could do it in a separate PR.